liquidctl/driver/nzxt_smart_device.py

```diff
@@ -182,66 +276,244 @@ def get_status(self, **kwargs):
+            self._write([0x22, 0x10, cid+1, 0x00] + leds[0:60]) # send first 20 colors to device (3 bytes per color)
+            self._write([0x22, 0x11, cid+1, 0x00] + leds[60:])  # send remaining colors to device
+            msg = self.device.read(self._READ_LENGTH) # wait for one reply before issuing command to specify color mode
+            LOGGER.debug('received %s', ' '.join(format(i, '02x') for i in msg))
+            self._write([0x22, 0xA0, cid+1, 0x00, mval, mod3, 0x00, channel_mod, 0x00,
+                         0x00, 0x64, 0x00, 0x32, 0x00, 0x00, 0x01])
+        elif mode == 'wings':  # wings requires special handling
+            for [g, r, b] in colors:
+                self._write([0x22, 0x10, cid+1])  # clear out all independent LEDs
+                self._write([0x22, 0x11, cid+1])  # clear out all independent LEDs
+                color_list = [g, r, b] * 8
+                r2 = (int)(r // 2.5)
+                g2 = (int)(g // 2.5)
+                b2 = (int)(b // 2.5)
+                color_list2 = [g2, r2, b2] * 8
+                r3 = (int)(r2 // 4)
+                g3 = (int)(g2 // 4)
+                b3 = (int)(b2 // 4)
+                color_list3 = [g3, r3, b3] * 8
+                for i in range(8):   #  send color scheme first, before enabling wings mode
+                    if i == 3 or i == 7:
+                        mod = 0x05
+                    else:
+                        mod = 0x01
+                    msg = ([0x22, 0x20, cid+1, i, 0x04, 0x39, 0x00, mod,
+                            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06,
+                            0x05, 0x85, 0x05, 0x85, 0x05, 0x85, 0x00, 0x00,
+                            0x00, 0x00, 0x00, 0x00])
+                    if i == 0 or i == 4:
+                        self._write(msg + color_list)
+                    elif i == 1 or i == 5:
+                        self._write(msg + color_list2)
+                    elif i == 2 or i == 6:
+                        self._write(msg + color_list3)
+                    else:
+                        self._write(msg)
+                self._write([0x22, 0x03, cid+1, 0x08])   # this actually enables wings mode
+        else:
+            channel_mod = [0x01, 0x20][cid]  # the purpose of this is unknown, but is based on cmd issued by CAM software
+            byte10 = 0x00
+            if mval == 0x03:  # for marquee-3|4|5|6 and backwards-marquee-3|4|5|6 we have to put led count in color_count
+                color_count = mod3
+                mod3 = 0
+            elif mval == 0x05:  # for all of the alternating, moving-alternating, and backwards-alternating modes
+                byte10 = mod3   # this is LED count (3 LEDs = 0, 4 LEDs = 1, 5 LEDs = 2, 6 LEDs = 3)
+                mod3 = 0x00
+                if (mod4 & 0x10) != 0:  # if 'moving' is chosen, set mod3 to 1
+                    mod3 = 0x01
+                mod4 = mod4 & 0x01  # mod4 represents 'backwards' -- need to strip off leading 1s
```

Comment:
Just let me see if I got this right... the message will resemble:

```python
self._write([0x28, 0x03, cid + 1, channel_mod, mval, sval,
                     either(0x00, 0x01), # 0x01 for moving, 0x00 otherwise
                     either(0x00, 0x01), # 0x01 for backwards, 0x00 otherwise
                     either(color_count, marquee_size),
                     either(0x00, alternating_size)]
            + list(itertools.chain(*colors)))
```

This hurts my brain.  Again, I can't emphasize this enough: I'm not blaming _you_ for my headache.

But maybe you can simplify a bit the code, so it will be less painful to maintain this for the future ourselves...  Basically, don't reassign to `mod3` and `mod4`, and instead explicitly name and compute each necessary value.

EDIT: the above is just pseudo-code to annotate the use of each byte.  I actually like that the header is computed before the write.  I really just think it would help to have more specific and non-overriding names for the those last 4 bytes of the header and, with that, simplify the earlier conditionals.

Reply:
Sure, this can be made more logical using general-purpose names in the header rather than specific-purpose names such as `color_count` which ends up having a particle/wave duality.

I'll work on this.